go/arrow/compute/internal/kernels/vector_run_end_encode.go

```diff
@@ -0,0 +1,947 @@
+		value, valid := re.readValue(readOffset)
+
+		if valid != curRunValid || !bytes.Equal(value, currentRun) {
+			writeValue(curRunValid, currentRun)
+			runEnd := R(readOffset - re.inputOffset)
+			outputRunEnds[writeOffset] = runEnd
+			writeOffset++
+			curRunValid, currentRun = valid, value
+		}
+
+		readOffset++
+	}
+
+	writeValue(curRunValid, currentRun)
+	outputRunEnds[writeOffset] = R(re.inputLen)
+	return writeOffset + 1
+}
+
+type runEndEncodeLoopBinary[R RunEndsType, O int32 | int64] struct {
+	inputLen, inputOffset      int64
+	inputValidity, inputValues []byte
+	offsetValues               []O
+	valueType                  arrow.DataType
+
+	estimatedValuesLen int64
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) readValue(idx int64) ([]byte, bool) {
+	if len(re.inputValidity) > 0 && bitutil.BitIsNotSet(re.inputValidity, int(idx)) {
+		return nil, false
+	}
+
+	start, end := re.offsetValues[idx], re.offsetValues[idx+1]
+	return re.inputValues[start:end], true
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) CountNumberOfRuns() (numValid, numOutput int64) {
+	re.estimatedValuesLen = 0
+
+	var offset int64
+	currentRun, curRunValid := re.readValue(offset)
+	offset++
+
+	if curRunValid {
+		numValid = 1
+		re.estimatedValuesLen += int64(len(currentRun))
+	}
+	numOutput = 1
+
+	for offset < re.inputLen {
+		value, valid := re.readValue(offset)
+		offset++
+		// new run
+		if valid != curRunValid || !bytes.Equal(value, currentRun) {
+			if valid {
+				re.estimatedValuesLen += int64(len(value))
+			}
+
+			currentRun = value
+			curRunValid = valid
+
+			numOutput++
+			if valid {
+				numValid++
+			}
+		}
+	}
+	return
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) PreallocOutput(ctx *exec.KernelCtx, numOutput int64, out *exec.ExecResult) {
+	runEndsBuffer := ctx.Allocate(int(numOutput) * int(SizeOf[R]()))
+	var validityBuffer *memory.Buffer
+	if len(re.inputValidity) > 0 {
+		validityBuffer = ctx.AllocateBitmap(numOutput)
+	}
+
+	valueBuffer := ctx.Allocate(int(re.estimatedValuesLen))
+	offsetsBuffer := ctx.Allocate(int(numOutput+1) * int(SizeOf[O]()))
+
+	reeType := arrow.RunEndEncodedOf(exec.GetDataType[R](), re.valueType)
+	*out = exec.ExecResult{
+		Type:   reeType,
+		Len:    re.inputLen,
+		Nulls:  0,
+		Offset: 0,
+		Children: []exec.ArraySpan{
+			{
+				Type: reeType.RunEnds(),
+				Len:  numOutput,
+			},
+			{
+				Type: reeType.Encoded(),
+				Len:  numOutput,
+			},
+		},
+	}
+
+	out.Children[0].Buffers[1].WrapBuffer(runEndsBuffer)
+	if validityBuffer != nil {
+		out.Children[1].Buffers[0].WrapBuffer(validityBuffer)
+	}
+	out.Children[1].Buffers[1].WrapBuffer(offsetsBuffer)
+	out.Children[1].Buffers[2].WrapBuffer(valueBuffer)
+}
+
+func (re *runEndEncodeLoopBinary[R, O]) WriteEncodedRuns(out *exec.ExecResult) int64 {
+	outputRunEnds := exec.GetData[R](out.Children[0].Buffers[1].Buf)
+	outputOffsets := exec.GetSpanOffsets[O](&out.Children[1], 1)
+	outputValues := out.Children[1].Buffers[2].Buf
+
+	var readOffset int64
```

Comment:
This is starting from `0` when it should start from the offset of the input.

Reply:
Not quite. the slice of the offsets `re.offsetValues` is already offset based on the offset of the input because `exec.GetSpanOffsets` which is called in `createVarBinaryEncoder` will apply the internal offset itself.